docs/content/sips/008-using-oci-registries.md

```diff
@@ -0,0 +1,298 @@
+  existing services - such as GitHub Container Registry, Docker Hub, AWS Elastic
+  Container Registry, Azure Container Registry, Google Artifact registry, or
+  others.
+- if using a managed service, scaling the infrastructure is no longer a concern
+  for users; if users decide to self-host, horizontally scaling a container
+  registry should be a more straightforward task, with more resources and
+  projects whose goal is scaling available when compared to Bindle.
+
+### The user experience
+
+The implementation should give users the ability to push an application to a
+compatible registry, pull an application locally, and run an application:
+
+```bash
+$ spin oci push ghcr.io/<username>/my-spin-application:v1
+INFO spin_publish::oci::client: Pushed "https://ghcr.io/v2/<username>/my-spin-application/manifests/sha256:9f4e7eebb27c0174fe6654ef5e0f908f1edc8f625324a9f49967ccde44a6516b"
+
+$ spin oci pull ghcr.io/<username>/my-spin-application:v1
+INFO spin_publish::oci::client: Pulled ghcr.io/<username>/my-spin-application:v1@sha256:9f4e7eebb27c0174fe6654ef5e0f908f1edc8f625324a9f49967ccde44a6516b
+
+$ spin oci run ghcr.io/radu-matei/my-spin-application:v1
+INFO spin_publish::oci::client: Pulled ghcr.io/<username>/my-spin-application:v1@sha256:9f4e7eebb27c0174fe6654ef5e0f908f1edc8f625324a9f49967ccde44a6516b
+Serving http://127.0.0.1:3000
+```
+
+The commands and arguments shown above are a suggestion - eventually, the
+`spin oci run` functionality should be integrated into `spin up`.
+
+### Authentication
+
+Historically, the `docker` CLI has been the toolchain used to interact with
+container images and registries - as a result, given its popularity, the
+`spin oci` functionality should be able to re-use credentials for already
+logged-in users. Additionally, most contianer registry services have
+instructions on how to log in to their services using the `docker login`
+command.
+
+However, having Docker installed locally should not be a prerequisite for using
+Spin. To address this, a `spin oci login` command should be implemented that
+authenticates the Spin CLI to the desired registry instance:
+
+```bash
+$ spin oci login --username <username> --password <password>
+# OR
+$ echo $CONTAINER_REGISTRY_PASSWORD | spin oci login --username <username> --password-stdin
+```
+
+This user experience would mirror
+[the `docker login` command](https://docs.docker.com/engine/reference/commandline/login/).
+
+### Migrating applications from Bindle to an OCI registry
+
+Changing the distribution mechanism for Spin applications is a breaking change
+for the project - to address this, the change should be implemented gradually,
+giving users the tool to migrate an application from Bindle to an OCI registry.
+
+This would be done in two steps: pulling the application from Bindle, then
+pushing it to the desired OCI registry:
+
+```bash
+$ spin bindle pull --bindle-server <server> --bindle-username <username> --bindle-password <password> [bindle-name]
```

Comment:
If we want to do this then we need a better sense of how this fits together - what is the bindle pulled into, what is push pushing from?  Currently `oci push` appears to push from a local manifest, similar to `bindle push` - is the idea to reconstitute a local app from a bindle?  This feels a bit odd given that we are going between two content-addressed formats.  But I could be wrong

An alternative approach might be to create a `bindle2oci` Spin plugin that pulls down the bindle, massages it into OCI layers, and pushes that directly.  This might also, longer term, help with migrating applications in the cloud, where we would want to do a migrate apps internally without a need for users to pull and push.

It might also be good to get a sense for how many users are affected by this - how many people are running apps in the Platform and would not be willing/able to redeploy from source?  That is, to what extent does this need to be a user tool vs a cloud internal one?

Reply:
We have used Bindle as the default mechanism for distributing Spin applications since the very first launch — because of that, I would be inclined to have this tool public.

The point re: this being a plugin rather than something built in is excellent — it makes so much more sense to pull that as it is needed. Great suggestion!

On pulling from Bindle, my assumption had been that it was easier to fetch a standalone bindle and push that, but I might be wrong. The two step approach here was based on that — happy to consider a single step approach that just translates the two formats.

TL; DR for me here — I feel strongly that, given that we will surely need to use this functionality, making it available to everyone makes a lot of sense. Building it as a plugin is a perfect use for a plugin, and I will update the SIP to reflect that.

In that case, the experience would be something like:

```
spin bindle2oci --bindle-server <server> --bindle-username <username> --bindle-password <password> ---bindle-name <name> --oci <new-reference>
```

How does that sound?

Reply:
Sounds good!

Reply:
Thanks for the suggestion, updated the SIP.